_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+expressions on par with other operators.
+
+In most cases built-in operators are only defined for [value types](https://docs.soliditylang.org/en/latest/types.html#value-types).
+Notable exceptions are the simple assignment and the "miscellaneous" group.
+For example, you can put almost anything in parentheses or branches of the ternary operator, as a part
+of a bigger expression, even if its type cannot be named or even assigned to anything.
+
+#### `using for`
+Since operators are a purely syntactic element, you can always use ordinary functions as a substitute.
+Solidity has a mechanism that is often used this way:
+[`using for`](https://docs.soliditylang.org/en/latest/contracts.html#using-for).
+
+The directive `using ... for ...` can be used to attach functions to a type.
+These functions can then be used as members on any object of that type and receive the object as
+their first parameter.
+
+```solidity
+pragma solidity ^0.8.13;
+
+type Int is int;
+using {add} for Int global;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function test(Int a, Int b) pure returns (Int) {
+    return a.add(b);
+}
+```
+
+Currently, `using for` has 3 forms:
+1. `using L for T`: attaches all non-private functions from library `L` to type `T`.
+1. `using L for *`: attaches all non-private functions from library `L` to any type, including types that
+    cannot be named in the language and used explicitly (e.g. types of literals or array slices).
+1. `using {f, L.g, ...} for T`: attaches given free or library functions `f`, `L.g`, ... to type `T`.
+
+The directive can be used:
+- ✅ At the file level.
+- ✅ Inside contract and library definitions.
+
+Can not be used:
+- ❌ Inside interface definitions.
+
+When used inside contracts, it is not inherited.
+Its effects extend only to the base contract that has it and not to any other contracts derived from it.
+
+At the file level, the directive affects all the free functions, contracts and libraries defined within that file.
+It does not affect other files that import the file containing it, unless it is marked as `global`.
+`global` extends its effect to all files where the type is visible.
+The use of `global` is possible only with UDVTs, structs, and enums that are defined in its source unit.
+
+### Introducing User-Defined Operators
+The new feature extends the third form of the `using for` syntax to allow binding operators to functions.
+The function is invoked whenever the operator is used on the values of the type:
+
+```solidity
+pragma solidity ^0.8.19;
+
+type Int is int;
+using {add as +} for Int global;
+
+function add(Int a, Int b) pure returns (Int) {
+    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
+}
+
+function test(Int a, Int b) pure returns (Int) {
+    return a + b; // Equivalent to add(a, b)
+}
+```
+
+An operator can only be bound in a `global` directive and the definition must be a `pure` free function.
```

Comment:
Here it would actually be nice to justify this a bit more. I.e. the primary use case for this is type libraries that build up a complete type that should always work the same everywhere and we don't have meanings of operators to depend on more than the operator types (i.e. not on scopes, etc.) to avoid pitfalls. Something along those lines :-).

Reply:
Yeah, I go into that more in the section on auditing. I generally tried to avoid too much detail in the intro section and initially had that in more specific sections instead. Now those sections are mostly gone though, as the one on auditing turned out to be a good place to list all the important corner cases in a natural way.